### PR TITLE
LUCENE-9068: Build FuzzyQuery automata up-front

### PR DESCRIPTION
**Pull request by romseygeek**
FuzzyTermsEnum can now either take an array of compiled automata, and 
an AttributeSource, to be used across multiple segments (eg during 
FuzzyQuery rewrite); or it can take a term, edit distance, prefix and transition
boolean and build the automata itself if only being used once (eg for fuzzy
nearest neighbour calculations).

Rather than interact via attribute sources and specialized attributes, users of
FuzzyTermsEnum can get the boost and set minimum competitive boosts
directly on the enum.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java [hunk 1]
```diff
@@ -58,30 +60,42 @@
...
+   * length <code>prefixLength</code> with <code>term</code> and which have at most {@code maxEdits} edits.
+   * <p>
+   * After calling the constructor the enumeration is already pointing to the first
+   * valid term if such a term exists.
+   *
+   * @param terms Delivers terms.
+   * @param term Pattern term.
+   * @param maxEdits Maximum edit distance.
+   * @param prefixLength the length of the required common prefix
+   * @param transitions whether transitions should count as a single edit
```
Comment (madrob) [inline lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: nit: why was this renamed from transpositions?

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java [hunk 2]
```diff
@@ -344,67 +363,4 @@ public BytesRef term() throws IOException {
...
-      }
-    }
-
-    @Override
-    public boolean equals(Object other) {
-      if (this == other)
-        return true;
-      if (!(other instanceof LevenshteinAutomataAttributeImpl))
-        return false;
-      return Arrays.equals(automata, ((LevenshteinAutomataAttributeImpl) other).automata);
```
Comment (madrob) [inline lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: nit: we have some unused imports after this deletion.

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java [hunk 3]
```diff
@@ -91,58 +105,49 @@
...
+      final int maxEdits, CompiledAutomaton[] automata) throws IOException {
+
     this.maxEdits = maxEdits;
     this.terms = terms;
     this.term = term;
+    this.atts = atts;
     
     // convert the string into a utf32 int[] representation for fast comparisons
-    this.termText = stringToUTF32(term.text());
+    int[] termText = stringToUTF32(term.text());
```
Comment (madrob) [inline lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: This seems like a potentially expensive conversion to have to do just to get the length when we already have the automata (and we might have already done it once, if this constructor was called via the other one on L96.)

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java (hunk 1 above)
Comment (romseygeek) [reply lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: Because I can't read or type, apparently... have fixed

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java (hunk 2 above)
Comment (romseygeek) [reply lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: fixed

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java (hunk 3 above)
Comment (romseygeek) [reply lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: I've rearranged things a bit now so that we don't call this twice

**Comment by romseygeek** [comment]
Thanks for the review @madrob, I've pushed an update based on your feedback and merged in the exception handling changes.  I've also done a bit more cleanup - `visit` for example now just delegates to the pre-built automata rather than trying to rebuild things.

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java [hunk 4]
```diff
@@ -92,76 +105,62 @@
...
-   * Builds a binary Automaton to match a fuzzy term
-   * @param text            the term to match
-   * @param prefixLength    length of a required common prefix
-   * @param transpositions  {@code true} if transpositions should count as a single edit
-   * @param maxEdits        the maximum edit distance of matching terms
+   * Sets the maximum non-competitive boost, which may allow switching to a
+   * lower max-edit automaton at run time
+   */
+  public void setMaxNonCompetitiveBoost(float boost) {
+    this.maxBoostAtt.setMaxNonCompetitiveBoost(boost);
```
Comment (madrob) [inline lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: Does this need to call `bottomChanged`?

File: lucene/core/src/test/org/apache/lucene/util/TestRamUsageEstimator.java [hunk 5]
```diff
@@ -32,6 +32,7 @@
 import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.DisjunctionMaxQuery;
 import org.apache.lucene.search.FuzzyQuery;
```
Comment (madrob) [inline lucene/core/src/test/org/apache/lucene/util/TestRamUsageEstimator.java]: This is an unused import failing recommit still.

File: lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java (hunk 4 above)
Comment (madrob) [reply lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: Thinking on this further... would it make sense to somehow check the terms before building all of the automata, and then possibly skip building the ed=2 case to begin with? Maybe checking the terms is more expensive than building the automaton, so we only care about saving time on how much we use the automaton at runtime? I'm not sure here and haven't done the profiling necessary to be confident.

Comment (romseygeek) [reply lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java]: > Does this need to call bottomChanged?

No, that's an internal call only, and it's delayed anyway so as to preserve term stats after calling `next()`.

> Maybe checking the terms is more expensive than building the automaton

I'd be surprised, unless the input term for the automaton is very long, but it's probably worth profiling and checking this in a followup.

**Comment by madrob** [comment]
LGTM after precommit comes back happy.
